File: src/lib/classes/Album.ts
import { Artist } from './Artist';
import { Copyright } from './Copyright';
import { ExternalId } from './ExternalId';
import { ExternalUrl } from './ExternalUrl';
import { SpotifyImage } from './SpotifyImage';
import type { Track } from './Track';

export class Album implements SpotifyData {
	// Saved Album
	added_at?: string;

	// Album Full
	artists?: Artist[];
	copyrights?: Copyright[];
	external_ids?: ExternalId;
	genres?: string[];
	popularity?: number;
	release_date?: string;
	release_date_precision?: string;
	total_tracks?: number;
	tracks: Track[] = [];

	// Album Simplified
	album_type!: string;
	available_markets?: string[];
	external_urls!: ExternalUrl;
	href!: string;
	id!: string;
	images!: SpotifyImage[];
	name!: string;
	type!: 'album';
	uri!: string;

	private constructor() {}

	public static fromSavedAlbum(saved: SpotifyApi.SavedAlbumObject, tracks: Track[]): Album {
		const album = Album.fromFullAlbum(saved.album, tracks);
		album.added_at = saved.added_at;
		return album;
	}

	public static fromFullAlbum(full: SpotifyApi.AlbumObjectFull, tracks: Track[]): Album {
		const album = Album.fromSimplifiedAlbum(full);
		album.artists = full.artists.map((a) => Artist.fromSimplifiedArtist(a));
		album.copyrights = full.copyrights.map((c) => Copyright.fromCopyrightObject(c));
		album.external_ids = ExternalId.fromExternalIdObject(full.external_ids);
		album.genres = full.genres;
		album.popularity = full.popularity;
		album.release_date = full.release_date;
		album.release_date_precision = full.release_date_precision;
		album.total_tracks = full.tracks.total;
		album.tracks = tracks;
		return album;
	}

	public static fromSimplifiedAlbum(simplified: SpotifyApi.AlbumObjectSimplified): Album {
		const album = new Album();
		album.album_type = simplified.album_type;
		album.available_markets = simplified.available_markets;
		album.external_urls = ExternalUrl.fromExternalUrlObject(simplified.external_urls);
		album.href = simplified.href;
		album.id = simplified.id;
		album.images = simplified.images.map((i) => SpotifyImage.fromImageObject(i));
		album.name = simplified.name;
		album.type = simplified.type;
		album.uri = simplified.uri;
		return album;
	}
}

File: src/main/java/de/wittenbude/expandify/models/spotifydata/Album.java
package de.wittenbude.expandify.models.spotifydata;

import com.neovisionaries.i18n.CountryCode;
import de.wittenbude.expandify.models.spotifydata.helper.Copyright;
import de.wittenbude.expandify.models.spotifydata.helper.Image;
import lombok.*;
import org.springframework.data.annotation.Id;
import org.springframework.data.mongodb.core.mapping.Document;
import org.springframework.data.mongodb.core.mapping.DocumentReference;
import se.michaelthelin.spotify.enums.AlbumType;
import se.michaelthelin.spotify.enums.ModelObjectType;
import se.michaelthelin.spotify.enums.ReleaseDatePrecision;

import java.util.Arrays;
import java.util.List;
import java.util.Map;

@Data
@NoArgsConstructor
@Document(collection = "albums")
public class Album {
    private AlbumType albumType;
    @DocumentReference(lazy = true)
    private List<ArtistSimplified> artists;
    private CountryCode[] availableMarkets;
    private Copyright[] copyrights;
    private Map<String, String> externalIds;
    private Map<String, String> externalUrls;
    private String[] genres;
    private String href;
    @Id
    private String id;
    private Image[] images;
    private String label;
    private String name;
    private Integer popularity;
    private String releaseDate;
    private ReleaseDatePrecision releaseDatePrecision;
    @DocumentReference(lazy = true)
    private List<TrackSimplified> tracks;
    private ModelObjectType type;
    private String uri;

    public Album(se.michaelthelin.spotify.model_objects.specification.Album album, List<TrackSimplified> tracks) {
        this.albumType = album.getAlbumType();
        this.artists = Arrays.stream(album.getArtists()).map(ArtistSimplified::new).toList();
        this.availableMarkets = album.getAvailableMarkets();
        this.copyrights = Arrays.stream(album.getCopyrights()).map(Copyright::new).toArray(Copyright[]::new);
        this.externalIds = album.getExternalIds().getExternalIds();
        this.externalUrls = album.getExternalUrls().getExternalUrls();
        this.genres = album.getGenres();
        this.href = album.getHref();
        this.id = album.getId();
        this.images = Arrays.stream(album.getImages()).map(Image::new).toArray(Image[]::new);
        this.label = album.getLabel();
        this.name = album.getName();
        this.popularity = album.getPopularity();
        this.releaseDate = album.getReleaseDate();
        this.releaseDatePrecision = album.getReleaseDatePrecision();
        this.type = album.getType();
        this.uri = album.getUri();
        this.tracks = tracks;
    }

}
